## Predicting ad positioning using Q-Learning

Placement of ads on website is the primary problem for companies that operate on ad revenue. The position where the ad is placed plays pivotal role on whether or not the ad will be clicked. Here we have the following choices:
	1. Place them randomly, or
	2. Place the ad on the same position

The problem with placing the ad on the same position is the user, after a certain time, will start ignoring the space since he's used to seeing ad at the place, he will end up ignoring that particular position hereafter. Hence, this will reduce the number of clicks on ads. The problem with the former option, placing them randomly, is it wouldn't take optimal positions into consideration. For instance, text beside images are viewed higher number of times than those text which are placed at a distance. It is infeasible to go through every website and repeat the procedure. 

Solution: Reinforcement Learning
Using Reinforcement Learning we can approximate the human behavior. 

Why Reinforcement Learning? 
We cannot use traditional Machine Learning here, since it requires:
	1. Huge data
	2. Features
	3. Tuning of many hyperparameters
And we neither have huge data, nor features. The only data we have is the position of the baner/ad and whether or not it was clicked. We will use this dataset from Kaggle: https://www.kaggle.com/akram24/ads-ctr-optimisation. We will solve this problem using Q-Learning. The reason for using Q-Learning here is :
	1. It is model free, so it doesn't require to know all the states.
	2. Intuitive to understand, and converges faster. 

In [1]:
# import routines
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import random

Our environment will be the dataset. It contains 10 ads position per row having values either 1, when the ad is clicked, or 0 when it is not. Every row can be considered as a state in the space, considering it kind of a navigation across multiple pages (on website, for instance) Lets load the dataset and visualize the first few rows

In [2]:
env = pd.read_csv('Ads_CTR_Optimisation.csv')
env.head()

Ad 1  Ad 2  Ad 3  Ad 4  Ad 5  Ad 6  Ad 7  Ad 8  Ad 9  Ad 10
0     1     0     0     0     1     0     0     0     1      0
1     0     0     0     0     0     0     0     0     1      0
2     0     0     0     0     0     0     0     0     0      0
3     0     1     0     0     0     0     0     1     0      0
4     0     0     0     0     0     0     0     0     0      0

## Random policy

If we were to not have Q-Learning, we would place the ads randomly at given positions. We will now simulate the same.  

In [3]:
# total rewards earned
reward = 0
# random policy: for every state, choose a random
# position for displaying the ad
for x in range(len(env)):
    action = np.random.randint(0, 10)
    # if the guess was correct, increase the reward
    if env.values[x][action] == 1:
        reward += 1
print("Reward collected: {}".format(reward))

Reward collected: 1245


## Using Max Policy
Another question we might ask, is to display the ad where it is clicked the most number of times. For instance, there might be a certain position where the ad clicked with a higher probability. Since the values of the rows is either 1 or 0, we can sum across the columns and count the number of times ad in the position was clicked. 

In [4]:
clicked_counts = env.values.sum(axis=0)
counts = pd.DataFrame({"ad": np.arange(1, 11), "counts": clicked_counts})
counts.set_index("ad")

counts
ad        
1     1703
2     1295
3      728
4     1196
5     2695
6      126
7     1112
8     2091
9      952
10     489

Which indicates ad 5 was clicked 2695 times. So if we were to always place an ad on position 5, it would be click around 2695 times. But can we do better? 

## Using Policy Iteration (Dynamic Programming) 

In [32]:
action_space = np.arange(0, 100)
# starting with random policy, choose a random choice for
# every state in the environment
state_size = len(env)
policy = [random.choice(action_space) for x in range(state_size)]
# will take random action for the first time
first_time = True
small_change = 1e-20
gamma = 0.9
episodes = 0
max_episodes = 10

V = dict()
# last positions reward will be 1
V[10000] = 1

# initially the value function for all states
# will be random values close to zero
for i in range(state_size):
    V[i] = np.random.random()

while episodes < max_episodes:
    # policy evaluation
    while True:
        if episodes > max_episodes:
            break
        episodes += 1
        if episodes % 100 == 0:
            print("Current episode: {}".format(episodes))
        biggest_change = 0
        # loop through every state present
        for state in range(state_size):
            old_V = V[state]
            # take random action according to policy
            action = policy[state]
            new_state = state + 1
            reward = env.values[state][action]
            V[state] = reward + gamma * V[new_state]
            biggest_change = max(biggest_change, abs(V[state] - old_V))
        if biggest_change < small_change:
            break
            
    # policy improvement
    policy_changed = False
    for state in range(state_size):
        best_val = -np.inf
        best_action = -1
        for action in action_space:
            new_state = state + 1
            reward = env.values[state][action]
            future_reward = reward + gamma * V[new_state]
            if future_reward > best_val:
                best_val = future_reward
                best_action = action
        assert best_action != -1
        if policy[state] != best_action:
            policy_changed = True
        policy[state] = best_action

    if not policy_changed:
        break
print("Total episodes trained: {}".format(episodes))

Total episodes trained: 11


In [33]:
# total rewards earned
reward = 0
# random policy: for every state, choose a random
# position for displaying the ad
for x in range(len(env)):
    action = policy[x]
    # if the guess was correct, increase the reward
    if env.values[x][action] == 1:
        reward += 1
print("Reward collected: {}".format(reward))

Reward collected: 10000


## Using Q-Learning



In [8]:
# using q-learning
states = len(env)
actions = 10
q_table = np.zeros((states, actions))

learning_rate = 0.7
# gamma = 0.618
gamma = 0.9

epsilon = 1.0
max_epsilon = 1.0
min_epsilon = 0.01
decay_rate = 0.01
max_episodes = 100

In [9]:
def exploit(eps):
    """Randomizes a number to select
    whether or not to expolit"""
    return np.random.uniform() > eps

def random_action():
    return np.random.randint(0, 10)

In [10]:
reward = 0
for episode in range(max_episodes):
    epsilon *= 2
    if episode % 50 == 0:
        print("Episode: {}".format(episode))
    for state in range(states):
        if exploit(epsilon):
            action = random_action()
        else:
            action = np.argmax(q_table[state])
        r = env.values[state][action]
        reward += r
        q_table[state][action] += learning_rate*(r + gamma*np.max(q_table[state, :]) - q_table[state][action])
    epsilon = min_epsilon + (max_epsilon - min_epsilon) * np.exp(-decay_rate*episode)

Episode: 0
Episode: 50


In [11]:
# test
rewards = 0
for state in range(states):
    best_action = np.argmax(q_table[state, :])
    r = env.values[state][best_action]
    rewards += r
print("Reward collected: {}".format(rewards))

Reward collected: 1703


In [16]:
env.values.sum(axis=1)

array([4, 2, 6, ..., 4, 6, 5])

In [30]:
rand_list = [np.random.randint(0, 10) for x in range(10000)]
env = np.zeros((10000, 100))
i = 0
for x in rand_list:
    env[i][x] = 1
    i += 1
env = pd.DataFrame(env)

In [31]:
env.sum().sum()

10000.0

In [18]:
env.values[0, :]

array([0, 1, 1, 0, 1, 0, 0, 0, 0, 1])